File: android/app/src/main/java/com/vocahq/vocaphone/core/GatewayEndpoint.kt
package com.vocahq.vocaphone.core

import java.net.URI

/**
 * VocaPhone is built around gateways the user runs themselves, so plain HTTP
 * has to stay available for LAN, mDNS and tailnet hosts. It is refused for hosts
 * that are reachable from the public internet, where an unencrypted bearer token
 * and transcript would be exposed.
 */
object GatewayEndpoint {

    sealed interface Validation {
        data class Valid(val url: String, val cleartext: Boolean) : Validation
        data class Invalid(val reason: String) : Validation
    }

    fun validate(value: String): Validation {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return Validation.Invalid("Enter your gateway address.")

        val uri = try {
            URI(trimmed)
        } catch (_: IllegalArgumentException) {
            return Validation.Invalid("This is not a valid address.")
        }

        val scheme = uri.scheme?.lowercase()
        if (scheme != "http" && scheme != "https") {
            return Validation.Invalid("The address must start with http:// or https://.")
        }
        val host = uri.host?.lowercase()
        if (host.isNullOrEmpty()) return Validation.Invalid("The address is missing a host name.")
        if (uri.userInfo != null) {
            return Validation.Invalid("Remove the user name from the address; the token authenticates you.")
        }
        if (!uri.query.isNullOrEmpty() || !uri.fragment.isNullOrEmpty()) {
            return Validation.Invalid("Remove the query or fragment from the address.")
        }

        val cleartext = scheme == "http"
        if (cleartext && !isPrivateHost(host)) {
            return Validation.Invalid(
                "$host is reachable from the public internet, so it needs https://."
            )
        }

        val port = if (uri.port == -1) "" else ":${uri.port}"
        val path = uri.path.orEmpty().trimEnd('/')
        return Validation.Valid("$scheme://$host$port$path", cleartext)
    }

    fun isCleartext(url: String): Boolean = url.startsWith("http://", ignoreCase = true)

    /**
     * True when traffic to [host] cannot leave a network the user controls: a
     * loopback or private address, an unqualified or mDNS name, or a Tailscale
     * MagicDNS name whose traffic is already inside a WireGuard tunnel.
     */
    fun isPrivateHost(host: String): Boolean {
        val name = host.trim('[', ']').lowercase().trimEnd('.')
        if (name.isEmpty()) return false
        if (name == "localhost") return true
        if (isPrivateIPv4(name)) return true
        if (isPrivateIPv6(name)) return true
        if (looksLikeIPv4(name) || name.contains(':')) return false
        if (name.endsWith(".local") || name.endsWith(".internal") || name.endsWith(".home.arpa")) {
            return true
        }
        // Tailscale MagicDNS names are fully qualified but only resolve, and only
        // carry traffic, inside the user's own tailnet.
        if (name == "ts.net" || name.endsWith(".ts.net")) return true
        // An unqualified single-label host can only be resolved by the local
        // network's own resolver.
        return !name.contains('.')
    }

    private fun looksLikeIPv4(host: String): Boolean =
        host.split('.').let { parts ->
            parts.size == 4 && parts.all { it.isNotEmpty() && it.all(Char::isDigit) }
        }

    private fun isPrivateIPv4(host: String): Boolean {
        if (!looksLikeIPv4(host)) return false
        val octets = host.split('.').map { it.toIntOrNull() ?: return false }
        if (octets.any { it !in 0..255 }) return false
        val (a, b) = octets[0] to octets[1]
        return when {
            a == 127 -> true                       // loopback
            a == 10 -> true                        // RFC 1918
            a == 172 && b in 16..31 -> true        // RFC 1918
            a == 192 && b == 168 -> true           // RFC 1918
            a == 169 && b == 254 -> true           // link-local
            a == 100 && b in 64..127 -> true       // CGNAT, used by Tailscale
            else -> false
        }
    }

    private fun isPrivateIPv6(host: String): Boolean {
        if (!host.contains(':')) return false
        val normalized = host.substringBefore('%')
        if (normalized == "::1") return true
        val leading = normalized.substringBefore(':').padStart(4, '0')
        val prefix = leading.take(2).toIntOrNull(16) ?: return false
        // fc00::/7 unique local, fe80::/10 link local.
        return prefix in 0xfc..0xfd || (prefix == 0xfe && leading[2] in '8'..'b')
    }
}

File: ios/VocaPhoneShared/GatewayEndpoint.swift
import Foundation

enum GatewayEndpoint {
    static func validatedURL(from value: String) -> URL? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: trimmed),
              let scheme = components.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              let host = components.host,
              !host.isEmpty,
              components.user == nil,
              components.password == nil,
              components.query == nil,
              components.fragment == nil
        else { return nil }

        components.scheme = scheme
        return components.url
    }

    static func usesUnencryptedHTTP(_ url: URL) -> Bool {
        url.scheme?.lowercased() == "http"
    }

    /// Cleartext pairing is restricted to networks the user controls. This
    /// matches the Android pairing contract; a bearer token must never be
    /// accepted from a QR code pointing to public HTTP.
    static func isPrivateNetworkHost(_ host: String) -> Bool {
        let name = host
            .trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
            .lowercased()
            .trimmingCharacters(in: CharacterSet(charactersIn: "."))
        guard !name.isEmpty else { return false }
        if name == "localhost" || isPrivateIPv4(name) || isPrivateIPv6(name) {
            return true
        }
        if looksLikeIPv4(name) || name.contains(":") { return false }
        if name.hasSuffix(".local") || name.hasSuffix(".internal") || name.hasSuffix(".home.arpa") {
            return true
        }
        if name == "ts.net" || name.hasSuffix(".ts.net") { return true }
        return !name.contains(".")
    }

    private static func looksLikeIPv4(_ host: String) -> Bool {
        let parts = host.split(separator: ".", omittingEmptySubsequences: false)
        return parts.count == 4 && parts.allSatisfy { !$0.isEmpty && $0.allSatisfy(\.isNumber) }
    }

    private static func isPrivateIPv4(_ host: String) -> Bool {
        guard looksLikeIPv4(host) else { return false }
        let octets = host.split(separator: ".").compactMap { Int($0) }
        guard octets.count == 4, octets.allSatisfy({ (0...255).contains($0) }) else { return false }
        switch (octets[0], octets[1]) {
        case (127, _), (10, _), (192, 168), (169, 254), (100, 64...127):
            return true
        case (172, 16...31):
            return true
        default:
            return false
        }
    }

    private static func isPrivateIPv6(_ host: String) -> Bool {
        let name = host.lowercased().split(separator: "%", maxSplits: 1).first ?? ""
        guard name.contains(":"), name != "::1" else { return name == "::1" }
        let firstGroup = name.split(separator: ":", omittingEmptySubsequences: false).first ?? ""
        let paddedGroup = String(repeating: "0", count: max(0, 4 - firstGroup.count))
            + String(firstGroup)
        guard let prefix = Int(paddedGroup, radix: 16) else { return false }
        return (0xfc00...0xfdff).contains(prefix) || (0xfe80...0xfebf).contains(prefix)
    }
}
